stdlib/test/builtin/test_string.mojo

```diff
@@ -593,12 +593,105 @@ fn test_rfind() raises:
     # assert_equal(String("🔥🔥").rfind("🔥"), 1)
 
 
+fn test_splitjoin() raises:
+    var text = String(
+        """Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi et ex leo.\
+        Vestibulum mollis, tellus nec pellentesque sagittis, massa ligula dapibus\
+        ipsum, in viverra orci odio ac neque. Aliquam ac velit nec sem interdum\
+        pellentesque at id erat. Nunc faucibus venenatis magna, sit amet \
+        vulputate erat viverra ac. Integer eget hendrerit nisi. Morbi efficitu\
+        , turpis a vulputate varius, libero libero gravida dui, quis dignissim\
+        turpis elit in turpis. Nulla cursus pharetra augue, in viverra felis \
+        tempor quis. Etiam commodo nisl a risus faucibus fringilla. Praesent \
+        pretium arcu tellus, non dignissim arcu efficitur et. Vestibulum accumsan\
+            bibendum libero, eu mollis lectus porta a.
+
+        Mauris nisl velit, blandit ut urna vel, cursus tempus lorem. Sed ac sodales libero. \
+        Donec pharetra consectetur lectus, nec imperdiet diam rhoncus in. Donec quis consequat \
+        felis, id egestas sem. Fusce tempus ex nisl, maximus commodo ante feugiat non. Proin \
+        viverra accumsan efficitur. Sed aliquam laoreet libero, et placerat felis venenatis \
+        quis. Cras ornare justo velit, luctus feugiat mauris ultricies vitae. Nullam augue \
+        urna, placerat sed tellus pharetra, convallis malesuada lectus. Suspendisse neque \
+        ligula, hendrerit dictum pellentesque sit amet, rhoncus sed diam. Ut augue mauris, \
+        """
+    )
+    for i in range(len(text)):
+        var sep = text[i]
+        if sep == " ":
+            continue
+        var joined = String("")
+        for it in text.split(sep):
+            joined += sep + it[]
+        assert_true(text == joined[len(sep) :])
+
+
+fn test_isspace_unicode() raises:
+    # test all utf8 and unicode separators
+    var _line_sep_utf8 = List[UInt8](0x20, 0x5C, 0x75, 0x32, 0x30, 0x32, 0x38)
+    """Unicode Line Separator: \\u2028."""
+    var _paragraph_sep_utf8 = List[UInt8](
+        0x20, 0x5C, 0x75, 0x32, 0x30, 0x32, 0x39
+    )
+    """Unicode Paragraph Separator: \\u2029."""
+    # TODO add line and paragraph separator as stringliteral once unicode
+    # escape secuences are accepted
+    var _universal_separators = List[String](
+        String("\n"),
+        String("\r"),
+        String("\v"),
+        String("\f"),
+        String("\x1c"),
+        String("\x1e"),
+        String("\x85"),
+        String(_line_sep_utf8),
+        String(_paragraph_sep_utf8),
+    )
+    for i in _universal_separators:
+        assert_true(i[].isspace())
+
+    for i in List[String]("not", "space", "", "s", "a", "c"):
+        assert_false(i[].isspace())
+
+
 fn test_split() raises:
```

Comment:
nice catch, fixed it now in the coming commit

Reply:
I think it might be better to use a `sep: Optional[String] = None` for the separator to denote no user inputed `sep` rather than a magic string. It matches the Python method signature better and allows `"a  b".split(None)` which while odd is a valid Python expression.

Reply:
I added it just to see if everything goes ok, the only problem with `sep: Optional[String] = None` is that it needs to be an `Optional[StringLiteral]` currently because it can't cast an embedded StringLiteral to a String implicitly yet.

Reply:
Adding a `sep: StringRef` overload solves that problem for now.

Reply:
As @bgreni suggested, adding an overload is the best we can do here (we have run into this issue before, it's a known limitation ATM). So you will basically end up with
```mojo
fn split(self, sep: StringRef, maxsplit: Int = -1) -> List[String]: ...
fn split(self, maxsplit: Int = -1) -> List[String]: ...  # delegate to the one above
```

Reply:
I think I just solved it with the trait I made WDYT? I'm working on solving the issue of `sep.join(some_str.split(sep)) == some_str` when sep=" ", there is an extra space in the original and not in the joined result for the test

Reply:
I think adding a trait for this might be overkill unless it could be useful otherwise?

Reply:
It's just a wrapper around `StringableCollectionElement, Sized, EqualityComparable` until we can do `fn split[T: StringableCollectionElement & Sized &  EqualityComparable]` or solve the problem of casting when inside an `Optional`